src/serac/physics/utilities/functional/domain_integral.hpp

```diff
@@ -227,212 +272,70 @@ void gradient_matrix_kernel(mfem::Vector& K_e, derivatives_type* derivatives_ptr
+
       // get the position of this quadrature point in the parent and physical space,
       // and calculate the measure of that point in physical space.
-      auto                    xi_q   = rule.points[q];
-      auto                    dxi_q  = rule.weights[q];
-      auto                    J_q    = make_tensor<dim, dim>([&](int i, int j) { return J(q, i, j, e); });
-      auto                    detJ_q = det(J_q);
-      [[maybe_unused]] double dx     = detJ_q * dxi_q;
+      auto xi_q  = rule.points[q];
+      auto dxi_q = rule.weights[q];
+      auto J_q   = make_tensor<dim, dim>([&](int i, int j) { return J(q, i, j, e); });
+      double dx = det(J_q) * dxi_q;
 
       // recall the derivative of the q-function w.r.t. its arguments at this quadrature point
+      //auto dq_darg = detail::AccessDerivatives(derivatives_ptr, e, q, rule, num_elements);
       auto dq_darg = detail::AccessDerivatives(derivatives_ptr, e, q, rule, num_elements);
 
-      // evaluate shape functions
-      [[maybe_unused]] auto M = test_element::shape_functions(xi_q);
-      [[maybe_unused]] auto N = trial_element::shape_functions(xi_q);
-      if constexpr (test_element::family == Family::HCURL) {
-        M = dot(M, inv(J_q));
-      }
-      if constexpr (trial_element::family == Family::HCURL) {
-        N = dot(N, inv(J_q));
-      }
-
-      [[maybe_unused]] auto f00 = serac::get<0>(serac::get<0>(dq_darg));
-      [[maybe_unused]] auto f01 = serac::get<1>(serac::get<0>(dq_darg));
-      [[maybe_unused]] auto f10 = serac::get<0>(serac::get<1>(dq_darg));
-      [[maybe_unused]] auto f11 = serac::get<1>(serac::get<1>(dq_darg));
-
-      // df0_du stiffness contribution
-      // size(M) = test_ndof
-      // size(N) = trial_ndof
-      // size(df0_du) = test_dim x trial_dim
-      if constexpr (!is_zero<decltype(f00)>::value) {
-        auto df0_du = convert_to_tensor_with_shape<test_dim, trial_dim>(f00);
-        for_loop<test_ndof, test_dim, trial_ndof, trial_dim>([&](auto i, auto id, auto j, auto jd) {
-          // maybe we should have a mapping for dofs x dim
-          K_elem[i + test_ndof * id][j + trial_ndof * jd] += M[i] * df0_du[id][jd] * N[j] * dx;
-        });
-      }
+      auto & q00 = serac::get<0>(serac::get<0>(dq_darg)); // derivative of source term w.r.t. field value
+      auto & q01 = serac::get<1>(serac::get<0>(dq_darg)); // derivative of source term w.r.t. field derivative
+      auto & q10 = serac::get<0>(serac::get<1>(dq_darg)); // derivative of   flux term w.r.t. field value
+      auto & q11 = serac::get<1>(serac::get<1>(dq_darg)); // derivative of   flux term w.r.t. field derivative
+
+      auto M = evaluate_shape_functions< test_element >(xi_q, J_q);
+      auto N = evaluate_shape_functions< trial_element >(xi_q, J_q);
+
+      // clang-format off
+      for (int i = 0; i < test_ndof; i++) {
+        for (int j = 0; j < trial_ndof; j++) {
+          K_elem[i][j] += (
```

Comment:
Nice. This is quiet a bit cleaner, although it seems to abstract how K_elem[i][j] is actually getting stored. Do we have control over that? (If only to match mfem assembled numbering).